libc/string_lenchr.c

```diff
@@ -150,9 +154,291 @@ TEST(string_len, big)
+}
+
+
+TEST(string_chr, torn)
+{
+	char *torn = "foo\0bar";
+	int sz;
+
+	/* Checking correct working of null terminating point */
+	sz = strlen(torn) + 1;
+	TEST_ASSERT_EQUAL_STRING("", strchr(torn, '\0'));
+	TEST_ASSERT_EQUAL_STRING("", strrchr(torn, '\0'));
+	TEST_ASSERT_EQUAL_STRING("", memchr(torn, '\0', sz));
+
+	TEST_ASSERT_EQUAL_STRING("foo", strchr(torn, 'f'));
+	TEST_ASSERT_EQUAL_STRING("foo", strrchr(torn, 'f'));
+	TEST_ASSERT_EQUAL_STRING("foo", memchr(torn, 'f', sz));
+
+	TEST_ASSERT_EQUAL_STRING(NULL, strchr(torn, 'b'));
+	TEST_ASSERT_EQUAL_STRING(NULL, strrchr(torn, 'b'));
+	TEST_ASSERT_EQUAL_STRING(NULL, memchr(torn, 'b', sz));
+}
+
+
+TEST(string_chr, whitespaces)
+{
+	const char whites[] = " \v\t\r\n";
+	int sz;
+
+	/* Checking if white signs doesn't make any interference on functions output */
+	sz = sizeof(whites) - 1;
+	TEST_ASSERT_EQUAL_STRING(&whites[sz - 1], strchr(whites, whites[sz - 1]));
+	TEST_ASSERT_EQUAL_STRING(&whites[sz - 1], strrchr(whites, whites[sz - 1]));
+	TEST_ASSERT_EQUAL_STRING(&whites[sz - 1], memchr(whites, whites[sz - 1], sz));
+
+	TEST_ASSERT_EQUAL_STRING(&whites[sz / 2], strchr(whites, whites[sz / 2]));
+	TEST_ASSERT_EQUAL_STRING(&whites[sz / 2], strrchr(whites, whites[sz / 2]));
+	TEST_ASSERT_EQUAL_STRING(&whites[sz / 2], memchr(whites, whites[sz / 2], sz));
+
+	TEST_ASSERT_EQUAL_STRING(&whites[0], strchr(whites, whites[0]));
+	TEST_ASSERT_EQUAL_STRING(&whites[0], strrchr(whites, whites[0]));
+	TEST_ASSERT_EQUAL_STRING(&whites[0], memchr(whites, whites[0], sz));
+
+	TEST_ASSERT_EQUAL_STRING(NULL, strchr(whites, 'x'));
+	TEST_ASSERT_EQUAL_STRING(NULL, strrchr(whites, 'x'));
+	TEST_ASSERT_EQUAL_STRING(NULL, memchr(whites, 'x', sz));
+}
+
+
+TEST(string_chr, empty)
+{
+	int sz;
+
+	/* Checking if we can get an empty string */
+	sz = 1;
+	TEST_ASSERT_EQUAL_STRING("", strchr("", '\0'));
+	TEST_ASSERT_EQUAL_STRING("", strrchr("", '\0'));
+	TEST_ASSERT_EQUAL_STRING("", memchr("", '\0', sz));
+
+	TEST_ASSERT_EQUAL_STRING("", strchr("", 0));
+	TEST_ASSERT_EQUAL_STRING("", strrchr("", 0));
+	TEST_ASSERT_EQUAL_STRING("", memchr("", 0, sz));
+
+	/* Looking for white spaces in empty string */
+	TEST_ASSERT_EQUAL_STRING(NULL, strchr("", ' '));
+	TEST_ASSERT_EQUAL_STRING(NULL, strrchr("", ' '));
+	TEST_ASSERT_EQUAL_STRING(NULL, memchr("", ' ', sz));
+
+	TEST_ASSERT_EQUAL_STRING(NULL, strchr("", '\n'));
+	TEST_ASSERT_EQUAL_STRING(NULL, strrchr("", '\n'));
+	TEST_ASSERT_EQUAL_STRING(NULL, memchr("", '\n', sz));
+
+	TEST_ASSERT_EQUAL_STRING(NULL, strchr("", '\t'));
+	TEST_ASSERT_EQUAL_STRING(NULL, strrchr("", '\t'));
+	TEST_ASSERT_EQUAL_STRING(NULL, memchr("", '\t', sz));
+
+	TEST_ASSERT_EQUAL_STRING(NULL, strchr("", '\r'));
+	TEST_ASSERT_EQUAL_STRING(NULL, strrchr("", '\r'));
+	TEST_ASSERT_EQUAL_STRING(NULL, memchr("", '\r', sz));
+
+	TEST_ASSERT_EQUAL_STRING(NULL, strchr("", '\v'));
+	TEST_ASSERT_EQUAL_STRING(NULL, strrchr("", '\v'));
+	TEST_ASSERT_EQUAL_STRING(NULL, memchr("", '\v', sz));
```

Comment:
Not necessarily needed, we have the edge cases like first character (and last in the same time - \0) and the character which does not exist in the string (space)

```suggestion
```